### ISSUE TITLE
Consider adding a means for annotations to require an extension

### ISSUE THREAD

**Issue by rherrmann** [junit-team/junit5#159]
A pattern that I see from time to time when using (JUnit 4) rules is that there is a _tagging_ annotation which is evaluated by a corresponding rule at runtime. For example [@RunInThread](https://gist.github.com/fappel/65982e5ea7a6b2fde5a3) or [@ConditionalIgnore](https://gist.github.com/rherrmann/7447571).

In both cases, annotating a test method without having the corresponding rule wouldn't make sense but there is nothing either to help the user discover this relation.

While playing around with JUnit 5 and expreimentally migrating those rules to use extensions, I thought it would be useful to be able to express that an annotation depends on a certain extension being present.

For example, if there was a `@RequiresExtension` annotation, the `DisabledWhen` annotation could express that it needs a `DisabledWhenExtension` to be present at runtime.
```Java

@RequiresExtension( DisabledWhenExtension.class )
@interface DisabledWhen {
  // ...
}

class SomeTest {
  @Test
  @DisabledWhen( ... )
  void disabledUnderCertainConditions() {
  }
}
```
... and when running the above test, the execution engine could let it fail with a clear error message.

Do you think this pattern is common enough to consider having a means to declare such dependency?

**Comment by jlink** [comment]
Maybe #121 could be a means to address this?

**Comment by rherrmann** [comment]
Yes, that seems like a perfect fit :-)

**Comment by sbrannen** [comment]
I don't think we need such a check.

The reason is that `@ExtendWith` can be used as a meta-annotation which in turn enables you to simply register the extension along with the annotation it acts on.

For example, you could write your extension like this:

```java
@ExtendWith(DisabledWhen.Extension.class)
public @interface DisabledWhen {

    static class Extension implements TestExecutionCondition {
    }

    // ...
}
```

With this feature of JUnit 5, it would be impossible to declare `@DisabledWhen` without having the appropriate extension or extensions registered.

In other words, your desire "_to be able to express that an annotation depends on a certain extension being present_" is already possible.

I am therefore closing this issue; however, if you feel that the above does not address your needs feel free to provide further input.

**Comment by rherrmann** [comment]
Thanks for the explanation. A truely elegant 'trick', and fully addresses my needs.

**Comment by sbrannen** [comment]
Great!

Thanks for the feedback.
